pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/dictionary/DictionaryIndexType.java

```diff
@@ -308,7 +321,20 @@ public static Dictionary read(PinotDataBuffer dataBuffer, ColumnMetadata metadat
             : new BigDecimalDictionary(dataBuffer, length, numBytesPerValue);
       case STRING:
         numBytesPerValue = metadata.getColumnMaxLength();
-        return loadOnHeap ? new OnHeapStringDictionary(dataBuffer, length, numBytesPerValue)
+
+        // If interning is enabled, get the required interners.
+        FALFInterner<String> strInterner = null;
+        FALFInterner<byte[]> byteInterner = null;
+        Intern internConfig = indexConfig.getIntern();
+        if (internConfig != null && !internConfig.isDisabled()) {
+          Preconditions.checkState(loadOnHeap, "Interning is only supported for on-heap dictionaries.");
+          DictionaryInternerHolder internerHolder = DictionaryInternerHolder.getInstance();
+          strInterner = internerHolder.getStrInterner(internIdentifierStr, internConfig.getCapacity());
+          byteInterner = internerHolder.getByteInterner(internIdentifierStr, internConfig.getCapacity());
+          LOGGER.info("Enabling interning for dictionary column: {}", columnName);
+        }
+
+        return loadOnHeap ? new OnHeapStringDictionary(dataBuffer, length, numBytesPerValue, strInterner, byteInterner)
```

Comment:
Should we not consider running the code between 325-335 only if the `loadOnHeap = true` ?

Reply:
Good point. Planned this change in my followup PR for leveraging interning with other onheap dictionaries.